contributors/design-proposals/scheduling/gang-scheduling.md

```diff
@@ -0,0 +1,188 @@
+
+## Function Detail
+
+### API Definition
+
+Although workload lifecycle requirements may be different, e.g. MPI vs. Spark, the requirements to scheduler are similar. To meet the requirements to scheduler, the following **Kind** is introduced in core part under `scheduling.k8s.io/v1alpha1` **Group**/**Version**.
+
+```go
+// PodSchedulingGroup defines the scheduling requirement of a pod group
+type PodSchedulingGroup struct {
+    metav1.TypeMeta
+    metav1.ObjectMeta
+
+    Spec PodSchedulingGroupTemplate
+    Status PodSchedulingGroupStatus
+}
+
+// LifecyclePolicy represents the lifecycle policy of PodSchedulingGroup
+// according to Pod's phase.
+type LifeCyclePolicy struct {
+    // The action that will be taken to the PodSchedulingGroup according to
+    // Pod's phase. One of "Restart", "None".
+    // Default to None.
+	Action Action
+
+    // The phase of pod; the controller takes actions according to this
+    // pod's phase. One of "PodFailed", "Unschedulable".
+    Event Event
+}
+
+// PodSchedulingGroupTemplate represents the template of a pod group.
+type PodSchedulingGroupTemplate struct {
+    // MinAvailable defines the minimal available tasks to run the Job;
+    // if there's not enough resources to start all tasks, the scheduler
+    // will not start anyone.
+    MinAvailable int
+
+    // Policy defines the policy of PodSchedulingGroup lifecycle.
+    // Default to 'Action: None, PodPhase: Failed'
+    // +optional
+    Policy []LifeCyclePolicy
+}
+
+// PodSchedulingGroupStatus represents the current state of a pod group.
+type PodSchedulingGroupStatus struct {
+    // The number of actively running pods.
+    // +optional
+    Running int32
+
+    // The number of pods which reached phase Succeeded.
+    // +optional
+    Succeeded int32
+
+    // The number of pods which reached phase Failed.
+    // +optional
+    Failed int32
+}
+```
+
+The `PodSchedulingGroup` , which is a namespaced object, specifies the scheduling constraints of a pod group, e.g. minimal available pods of a group, lifecycle policy. To define which pods are member of `PodSchedulingGroup`, the following field is introduced to `PodSpec`.
+
+```go
+// PodSpec is a description of a pod.
+type PodSpec struct {
+    ...
+
+    // The name of PodSchedulingGroup that the pod belongs to; the pod
+    // must belong to the PodSchedulingGroup in the same namespace.
+    // +optional
+    GroupName string
```

Comment:
I assume this will be an alpha field in the API?

Reply:
Yes.